contracts/ConnectivityTrackerHbbft.sol

```diff
@@ -0,0 +1,243 @@
+// SPDX-License-Identifier: Apache 2.0
+pragma solidity =0.8.17;
+
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {EnumerableSetUpgradeable} from "@openzeppelin/contracts-upgradeable/utils/structs/EnumerableSetUpgradeable.sol";
+
+import {IConnectivityTrackerHbbft} from "./interfaces/IConnectivityTrackerHbbft.sol";
+import {IValidatorSetHbbft} from "./interfaces/IValidatorSetHbbft.sol";
+import {IStakingHbbft} from "./interfaces/IStakingHbbft.sol";
+
+contract ConnectivityTrackerHbbft is
+    Initializable,
+    OwnableUpgradeable,
+    IConnectivityTrackerHbbft
+{
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+
+    IValidatorSetHbbft public validatorSetContract;
+    IStakingHbbft public stakingContract;
+
+    uint256 public minReportAgeBlocks;
+    uint256 public earlyEpochEndToleranceLevel;
+
+    mapping(uint256 => bool) public isEarlyEpochEnd;
+    mapping(uint256 => mapping(address => uint256)) public validatorConnectivityScore;
+
+    mapping(uint256 => EnumerableSetUpgradeable.AddressSet) private _flaggedValidators;
+    mapping(uint256 => mapping(address => EnumerableSetUpgradeable.AddressSet)) private _reporters;
+
+    event SetMinReportAgeBlocks(uint256 _minReportAge);
+    event SetEarlyEpochEndToleranceLevel(uint256 _level);
+    event ReportMissingConnectivity(
+        address indexed reporter,
+        address indexed validator,
+        uint256 indexed blockNumber,
+        uint256 blockHash
```

Comment:
while it is good to have the blockNumber as additional field in the event, so an observer can see the latency of the block,
we do not need the blockHash data here.